### DISCUSSION TITLE
add submodule support

### DISCUSSION THREAD

**Discussion by geiseri** [microsoft/vcpkg#20994] [Ideas]
Many projects use git submodules for support components.  While saying everything must be a vcpkg port this is not reasonable for dependencies that are specific versions or embedded components of a larger project. These dependencies are internal so the risk of conflicting versions is low.

**Comment by geiseri** [comment]
I guess I am wondering why there is such pushback from the maintainers.  This has been the subject of multiple requests and workarounds.  Is there a concrete case of why this is a bad idea?

**Comment by jasonbeverage** [comment]
I'm actually running into this now as well.  I'm trying to update vcpkg to build the latest osgEarth which uses submodules now for a few external dependencies and not finding a clean way to do it.   I'm just getting into this but is there pushback on projects that use submodules in vcpkg?  As you mentioned, the dependencies we are pulling down are internal and tied to specific versions so should be low risk to pull down as a submodule.

**Comment by Neumann-A** [comment]
You can always manually setup the download of the submodule within the portfile if there is no port for the dependency.

**Reply by geiseri** [reply]
It is not about the port though.  There is a reason that the upstream author has the component as a submodule vs an external dependency.  Sometimes it is an explicit version.  Sometimes it is out of convenience.  In the cases of convenience, those are usually configured time options to exclude.

**Comment by jasonbeverage** [comment]
I came across this [issue](https://github.com/microsoft/vcpkg/issues/6886) which does provide some solutions for using git manually.  

This seems to work but it's lots of extra code compared to the typical vcpkg_from_github or vcpkg_from_git.  Both vcpkg_from_github and vcpkg_from_git will do things like apply patches for you with a single command.

It looks like vcpkg_from_github just downloads the source ref as a tar.gz from Github, so you can't apply submodules there.  I suppose you could modify vcpkg_from_git to update the submodules optionally since it seems to be using git directly.

Are there other projects that use git submodules in vcpkg currently?

**Reply by Neumann-A** [reply]
> Are there other projects that use git submodules in vcpkg currently?

qttools, qtwebengine, itk, paraview

**Reply by geiseri** [reply]
Yes, you would think for all the engineering hours spent on working around this there would be something considered for the project.   The biggest issue I have with the current situation is best displayed by `paraview`.  The steps to maintaining the submodule references along with the upstream are arduous at best, or a source for bugs at worse.    I guess that does bring me back to the technical reasons as to why the authors of vcpkg are so resistant to git submodule support.

**Reply by jasonbeverage** [reply]
Thanks!   I'll take a look at those and use them for reference.

**Comment by JackBoosY** [comment]
Fixing this in https://github.com/microsoft/vcpkg/pull/21591.

**Reply by anthonyalayo** [reply]
this never got merged
